app/src/main/kotlin/com/simplemobiletools/camera/helpers/MediaOutputHelper.kt

```diff
@@ -0,0 +1,210 @@
+import com.simplemobiletools.commons.extensions.getDocumentFile
+import com.simplemobiletools.commons.extensions.getDoesFilePathExist
+import com.simplemobiletools.commons.extensions.getFileOutputStreamSync
+import com.simplemobiletools.commons.extensions.getFilenameFromPath
+import com.simplemobiletools.commons.extensions.getMimeType
+import com.simplemobiletools.commons.extensions.hasProperStoredAndroidTreeUri
+import com.simplemobiletools.commons.extensions.hasProperStoredFirstParentUri
+import com.simplemobiletools.commons.extensions.hasProperStoredTreeUri
+import com.simplemobiletools.commons.extensions.isAccessibleWithSAFSdk30
+import com.simplemobiletools.commons.extensions.isRestrictedSAFOnlyRoot
+import com.simplemobiletools.commons.extensions.needsStupidWritePermissions
+import com.simplemobiletools.commons.extensions.showFileCreateError
+import java.io.File
+import java.io.OutputStream
+
+class MediaOutputHelper(private val activity: BaseSimpleActivity) {
+
+    companion object {
+        private const val TAG = "MediaOutputHelper"
+        private const val MODE = "rw"
+    }
+
+    private val mediaStorageDir = activity.config.savePhotosFolder
+
+    fun getOutputStreamMediaOutput(): MediaOutput.OutputStreamMediaOutput? {
+        val canWrite = activity.canWrite(mediaStorageDir)
+        Log.i(TAG, "getMediaOutput: canWrite=${canWrite}")
+        return if (canWrite) {
+            val path = activity.getOutputMediaFile(true)
+            val uri = activity.getUri(path)
+            uri?.let {
+                activity.getFileOutputStreamSync(path, path.getMimeType())?.let {
+                    MediaOutput.OutputStreamMediaOutput(it, uri)
+                }
+            }
+        } else {
+            null
+        }.also {
+            Log.i(TAG, "output stream: $it")
+        }
+    }
+
+    fun getFileDescriptorMediaOutput(): MediaOutput.FileDescriptorMediaOutput? {
+        val canWrite = activity.canWrite(mediaStorageDir)
+        Log.i(TAG, "getMediaOutput: canWrite=${canWrite}")
+        return if (canWrite) {
+            val path = activity.getOutputMediaFile(false)
+            val uri = activity.getUri(path)
+            uri?.let {
+                activity.getFileDescriptorSync(path, path.getMimeType())?.let {
+                    MediaOutput.FileDescriptorMediaOutput(it, uri)
+                }
+            }
+        } else {
+            null
+        }.also {
+            Log.i(TAG, "descriptor: $it")
+        }
+    }
+
+    private fun BaseSimpleActivity.canWrite(path: String): Boolean {
+        return when {
+            isRestrictedSAFOnlyRoot(path) -> hasProperStoredAndroidTreeUri(path)
+            needsStupidWritePermissions(path) -> hasProperStoredTreeUri(false)
+            isAccessibleWithSAFSdk30(path) -> hasProperStoredFirstParentUri(path)
+            else -> File(path).canWrite()
+        }
+    }
+
+    private fun BaseSimpleActivity.getUri(path: String): Uri? {
```

Comment:
is this a private extension or what exactly? Looks weird.

Reply:
Fixed